typed/lib.clojure/src/typed/ann/clojure/base.cljc

```diff
@@ -0,0 +1,754 @@
+;   (defmulti g odd?)
+;
+; 2. what bound do we want on `f` and `d`?
+;   ; say if we don't have `d`'s lower bound as EveryIFn..
+;   (ann f (MultiFn .. ':a))
+;   (defmulti f :a)
+;
+;   (ann g (MultiFn .. (ToIFn ':a)))
+;   (defmulti g (fn [a] (:a a)))
+;
+;   ; is this desirable? let's use the lower bound, since we can always
+;   ; make the type more permissive later.
+;   (ann-form f (TypeOf g)) ;ok
+;   (ann-form g (TypeOf f)) ;fails
+
+#_#_
+MultiFn [[[f :variance :covariant :> EveryIFn :< AnyIFn]
+          [d :variance :covariant :> EveryIFn :< AnyIFn]
+          ;; only support :default default for now, and no hierarchy support
+          #_#_
+          :named
+          {default [:variance :covariant,
+                    :between [EveryIFn AnyValue],
+                    :default ':default]
+           hierarchy [:variance :covariant
+                      :< (Var Hierarchy)
+                      :default GlobalHierarchy]}]
+         :replace
+         {clojure.lang.IFn f}]
+
+;;; We override the internal Java classes that clojure.lang.* classes use
+;;; and simulate some of them extending Clojure interfaces as if they were protocols
+
+; Hack for Seqable things. Not needed if Seqable was a protocol.
+
+java.lang.CharSequence [[]
+                        :unchecked-ancestors
+                        [(Seqable (t/NilableNonEmptySeq Character))
+                         (Indexed Character)]]
+
+;FIXME Need to correctly check ancestors, this shouldn't be necessary because String is a CharSequence
+; CTYP-15
+java.lang.String [[]
+                  :replace
+                  {Comparable (Comparable String)}
+                  :unchecked-ancestors
+                  [(Seqable (t/NilableNonEmptySeq Character))
+                   (Indexed Character)]]
+
+java.lang.Iterable [[[a :variance :covariant]]
+                    :unchecked-ancestors
+                    [(Seqable (t/NilableNonEmptySeq a))]]
+
+;; the following Java collections should really have :invariant type params.
+;; Scala deals with this via implicit conversions from their own
+;; immutable versions of these interfaces.
+;; https://docs.scala-lang.org/overviews/collections/conversions-between-java-and-scala-collections.html
+;; We could do something similar by introducing similarly named covariant types that could
+;; coexist with the invariant versions. eg., (typed.clojure.java.immutable/Set x)
+java.util.Set [[[a :variance :covariant]]
+               :replace
+               {;Iterable (Iterable a)
+                Collection (Collection a)}
+               :unchecked-ancestors
+               [(Seqable (t/NilableNonEmptySeq a))]]
+
+;; FIXME should invariant
+java.util.Collection [[[a :variance :covariant]]
+                      :replace
+                      {Iterable (Iterable a)}
+                      :unchecked-ancestors
+                      [(Seqable (t/NilableNonEmptySeq a))]]
+
+java.lang.ref.Reference [[[a :variance :invariant]]]
```

Comment:
For class overrides, ensure we don't override JVM classes in CLR. Also ensure we add appropriate CLR type overrides for types we use as ancestors in this file.